stable/dex/templates/pre-upgrade-delete-jobs.yaml

```diff
@@ -0,0 +1,76 @@
+{{- if or .Values.certs.grpc.create .Values.certs.web.create }}
+# For upgrades to 2.12.3+.
+# Delete the Jobs created by the dex chart prior to upgrading. Priority class was added
+# to Job specs, which is an immutable field and requires the Job to be
+# deleted and recreated. After this release, we can remove this pre-upgrade hook because
+# TTL will be set on the Job to automatically clean it up after it runs.
+---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: {{ template "dex.fullname" . }}-pre-upgrade
+  namespace: {{ .Release.Namespace }}
+  annotations:
+    helm.sh/hook: pre-upgrade
+    helm.sh/hook-weight: "-5"
+    helm.sh/hook-delete-policy: hook-succeeded,before-hook-creation
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  name: {{ template "dex.fullname" . }}-pre-upgrade
+  namespace: {{ .Release.Namespace }}
+  annotations:
+    "helm.sh/hook": pre-upgrade
+    "helm.sh/hook-weight": "-4"
+    "helm.sh/hook-delete-policy": hook-succeeded,before-hook-creation
+rules:
+  - apiGroups: ["batch"]
+    resources: ["jobs"]
+    verbs: ["get", "watch", "list", "delete"]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: {{ template "dex.fullname" . }}-pre-upgrade
+  namespace: {{ .Release.Namespace }}
+  annotations:
+    "helm.sh/hook": pre-upgrade
+    "helm.sh/hook-weight": "-4"
+    "helm.sh/hook-delete-policy": hook-succeeded,before-hook-creation
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: {{ template "dex.fullname" . }}-pre-upgrade
+subjects:
+  - kind: ServiceAccount
+    name: {{ template "dex.fullname" . }}-pre-upgrade
+    namespace: {{ .Release.Namespace }}
+---
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: {{ template "dex.fullname" . }}-delete-jobs
+  namespace: {{ .Release.Namespace }}
+  annotations:
+    "helm.sh/hook": pre-upgrade
+    "helm.sh/hook-weight": "4"
+    "helm.sh/hook-delete-policy": hook-succeeded,before-hook-creation
+spec:
+  template:
+    metadata:
+      name: {{ template "dex.fullname" . }}-delete-jobs
+    spec:
+      serviceAccountName: {{ template "dex.fullname" . }}-pre-upgrade
+      restartPolicy: OnFailure
+      {{- if .Values.priorityClassName }}
+      priorityClassName: "{{ .Values.priorityClassName }}"
+      {{- end }}
+      containers:
+        - name: kubectl
+          image: "{{ .Values.kubectlImage }}"
+          command:
+            - sh
+            - -c
+            - kubectl delete jobs.batch -l 'app.kubernetes.io/component in (job-grpc-certs, job-web-certs),app.kubernetes.io/name=dex' --cascade=orphan -n {{ .Release.Namespace }}
```

Comment:
i'm not sure what this job exactly is used for but now that we are deleting it, i think it would be recreated and reran during the upgrade, right?

I'm wondering if rerunning it would cause any issues during upgrade 🤔. since it regenerates some certs, are there other clients that need to get the new CA or something?

Reply:
I tested this upgrade on the daily and didn't see any (obvious) issues with regenerating the certs, but I'd like @mhrabovcin or @palexster to confirm. Since this job really doesn't need to run on every upgrade though.. I will remove the TTL, and also move the pre-upgrade hook out of the chart and into k-apps, to limit it to only run during upgrade from 2.5 -> 2.6.

Reply:
dex helmrelease isn't behind a flux kustomization in k-apps, so upgrade will result in a reinstall 😬 should be a small amount of downtime in dex.. I don't think it will have significant impact.

Reply:
ok, went all over the place with this one 😓 but.. weighing all the things. I wanted to make a pre-upgrade hook here somewhat agnostic from DKP versioning but maybe it's fine, since this chart is really only used in the context of kommander. So to avoid the reinstall of dex (due to moving it behind a flux kustomization in k-apps), I can leave the pre-upgrade hook here with the version constraint to only delete the jobs if they were deployed as part of the 2.11.1 dex chart version (the version deployed in dkp 2.5). That way it only deletes the job for upgrades from dkp 2.5 to 2.6, and not within 2.6. Keeping it in the chart means that the priority of the pre-upgrade hook can also be configurable (vs putting it in a manifest in k-apps)

Reply:
done in https://github.com/mesosphere/charts/pull/1439/commits/d1c83bb945494e9673dc8a8418eb883402aa2b04

queuing up k-apps branch: https://github.com/mesosphere/kommander-applications/pull/1350

Reply:
sounds good to me!